src/Orleans/AssemblyLoader/AssemblyLoader.cs

```diff
@@ -3,104 +3,26 @@
 using System.IO;
 using System.Linq;
 using System.Reflection;
-using System.Text;
-using Orleans.CodeGeneration;
 
 namespace Orleans.Runtime
 {
-    internal class AssemblyLoader
+    public class AssemblyLoader
     {
-        private readonly Dictionary<string, SearchOption> dirEnumArgs;
-        private readonly HashSet<AssemblyLoaderPathNameCriterion> pathNameCriteria;
-        private readonly HashSet<AssemblyLoaderReflectionCriterion> reflectionCriteria;
+        public IAssemblyCatalog AssemblyCatalog { get; private set; }
         private readonly Logger logger;
-        internal bool SimulateExcludeCriteriaFailure { get; set; }
-        internal bool SimulateLoadCriteriaFailure { get; set; }
-        internal bool SimulateReflectionOnlyLoadFailure { get; set; }
-        internal bool RethrowDiscoveryExceptions { get; set; }
 
-        private AssemblyLoader(
-                Dictionary<string, SearchOption> dirEnumArgs,
-                HashSet<AssemblyLoaderPathNameCriterion> pathNameCriteria,
-                HashSet<AssemblyLoaderReflectionCriterion> reflectionCriteria,
-                Logger logger)
+        private AssemblyLoader(IAssemblyCatalog catalog)
         {
-            this.dirEnumArgs = dirEnumArgs;
-            this.pathNameCriteria = pathNameCriteria;
-            this.reflectionCriteria = reflectionCriteria;
-            this.logger = logger;
-            SimulateExcludeCriteriaFailure = false;
-            SimulateLoadCriteriaFailure = false;
-            SimulateReflectionOnlyLoadFailure = false;
-            RethrowDiscoveryExceptions = false;
+            AssemblyCatalog = catalog;
 
-            // Ensure that each assembly which is loaded is processed.
-            AssemblyProcessor.Initialize();
-
-            // initialize serialization for all assemblies to be loaded.
-            AppDomain.CurrentDomain.AssemblyLoad += OnAssemblyLoad;
-
-            Assembly[] assemblies = AppDomain.CurrentDomain.GetAssemblies();
-
-            // initialize serialization for already loaded assemblies.
-            CodeGeneratorManager.GenerateAndCacheCodeForAllAssemblies();
-            foreach (var assembly in assemblies)
-            {
-                AssemblyProcessor.ProcessAssembly(assembly);
-            }
-        }
-
-
-        /// <summary>
-        /// Handles <see cref="AppDomain.AssemblyLoad"/> events.
-        /// </summary>
-        /// <param name="sender">The sender of the event.</param>
-        /// <param name="args">The event arguments.</param>
-        private static void OnAssemblyLoad(object sender, AssemblyLoadEventArgs args)
-        {
-            AssemblyProcessor.ProcessAssembly(args.LoadedAssembly);
+            AssemblyProcessor.ProcessAssemblies(catalog.GetAssemblies());
```

Comment:
Is there a nice way that we can move this work out of the constructor and into another method?